File: python/src/yolo26_kit/core/decode_raw.py
"""Decoder for non-e2e raw YOLO26 ONNX exports (1, 4+nc, N).

See spec/decode.md Algorithm D.
"""
from __future__ import annotations

from collections.abc import Iterable
from typing import Literal, cast

import numpy as np

from ._axes import _split_channel_anchor_axes
from .types import COCO_CLASSES, Detection

_FormatT = Literal["dict", "arrays"]


def _sigmoid(x: np.ndarray) -> np.ndarray:
    return cast(np.ndarray, 1.0 / (1.0 + np.exp(-x)))


def decode_detect(
    output: np.ndarray,
    conf: float = 0.25,
    classes: Iterable[int] | None = None,
    min_area: float | None = None,
    format: _FormatT = "dict",
    *,
    num_classes: int | None = None,
    assume_sigmoid: bool = True,
    strict: bool = False,
    strict_dtype: bool = False,
) -> list[Detection] | dict[str, np.ndarray]:
    if not 0.0 <= conf <= 1.0:
        raise ValueError(f"conf must be in [0, 1]; got {conf}")
    arr = np.asarray(output)
    canonical, _ch = _split_channel_anchor_axes(arr, num_classes=num_classes)

    if arr.dtype != np.float32:
        if strict_dtype:
            raise ValueError(f"expected float32; got {arr.dtype}")
        canonical = canonical.astype(np.float32, copy=False)

    boxes_cxcywh = canonical[0, 0:4, :]
    cls = canonical[0, 4:, :]
    if not assume_sigmoid:
        cls = _sigmoid(cls)

    scores = cls.max(axis=0)
    classes_arr = cls.argmax(axis=0).astype(np.int32)

    cx, cy, w, h = boxes_cxcywh[0], boxes_cxcywh[1], boxes_cxcywh[2], boxes_cxcywh[3]
    x1 = cx - w * 0.5
    y1 = cy - h * 0.5
    x2 = cx + w * 0.5
    y2 = cy + h * 0.5
    boxes = np.stack([x1, y1, x2, y2], axis=1)

    finite = np.isfinite(scores) & np.isfinite(boxes).all(axis=1)
    if not finite.all() and strict:
        raise ValueError("output contains NaN/Inf")
    mask = finite & (scores >= conf)

    # Class-id range validation (argmax may pick a class index whose
    # underlying number of classes exceeds the COCO label table).
    valid_cls = (classes_arr >= 0) & (classes_arr < len(COCO_CLASSES))
    if not valid_cls.all():
        if strict:
            raise ValueError("class_id out of range")
        mask &= valid_cls

    if classes is not None:
        allowlist = np.fromiter(classes, dtype=np.int32)
        if allowlist.size:
            if (allowlist < 0).any() or (allowlist >= len(COCO_CLASSES)).any():
                raise ValueError(f"classes allowlist out of range: {allowlist.tolist()}")
            mask &= np.isin(classes_arr, allowlist)

    if min_area is not None:
        widths = boxes[:, 2] - boxes[:, 0]
        heights = boxes[:, 3] - boxes[:, 1]
        mask &= (widths * heights) >= min_area

    boxes = boxes[mask]
    scores = scores[mask]
    classes_arr = classes_arr[mask]

    src_idx = np.arange(scores.shape[0], dtype=np.int64)
    order = np.lexsort((src_idx, classes_arr, -scores))
    boxes = boxes[order]
    scores = scores[order]
    classes_arr = classes_arr[order]

    if format == "arrays":
        return {
            "boxes": np.ascontiguousarray(boxes, dtype=np.float32),
            "scores": np.ascontiguousarray(scores, dtype=np.float32),
            "classes": np.ascontiguousarray(classes_arr, dtype=np.int32),
        }
    if format == "dict":
        return [
            {
                "box": [float(b[0]), float(b[1]), float(b[2]), float(b[3])],
                "score": float(s),
                "class": int(c),
                "label": COCO_CLASSES[int(c)],
            }
            for b, s, c in zip(boxes, scores, classes_arr, strict=True)
        ]
    raise ValueError(f"unknown format: {format!r}")

File: js/src/core/decodeRaw.ts
import { splitChannelAnchorAxes } from "./_axes";
import { COCO_CLASSES } from "./classes";
import { classAwareNMS } from "./nms";
import type { ArrayDetections, DecodeRawOptions, Detection } from "./types";

const sigmoid = (x: number) => 1 / (1 + Math.exp(-x));

function readF32(arr: Float32Array, i: number): number {
  // biome-ignore lint/style/noNonNullAssertion: index bounds guaranteed by caller
  return arr[i]!;
}

export function decodeDetect(
  data: Float32Array,
  shape: readonly number[],
  opts: DecodeRawOptions = {},
): Detection[] | ArrayDetections {
  const {
    conf = 0.25,
    format = "dict",
    assumeSigmoid = true,
    strict = false,
    numClasses,
    classes,
    minArea,
    nms = true,
    iouThreshold = 0.45,
  } = opts;
  if (conf < 0 || conf > 1) throw new Error(`conf must be in [0, 1]; got ${conf}`);

  const { canonicalAxes } = splitChannelAnchorAxes(shape, numClasses);
  const { channelStride, anchorStride, channels, anchors } = canonicalAxes;
  const nc = channels - 4;

  const at = (ch: number, n: number) => readF32(data, ch * channelStride + n * anchorStride);

  const allowlist = classes ? new Set(classes) : null;
  if (allowlist) {
    for (const c of allowlist) {
      if (c < 0 || c >= COCO_CLASSES.length) {
        throw new Error(`classes allowlist out of range: ${c}`);
      }
    }
  }

  const boxesArr: number[] = [];
  const scoresArr: number[] = [];
  const clsArr: number[] = [];
  const idxs: number[] = [];

  for (let i = 0; i < anchors; i++) {
    const cx = at(0, i);
    const cy = at(1, i);
    const w = at(2, i);
    const h = at(3, i);
    let bestRaw = Number.NEGATIVE_INFINITY;
    let bestC = 0;
    for (let c = 0; c < nc; c++) {
      const v = at(4 + c, i);
      if (v > bestRaw) {
        bestRaw = v;
        bestC = c;
      }
    }
    const score = assumeSigmoid ? bestRaw : sigmoid(bestRaw);
    const finite =
      Number.isFinite(cx) &&
      Number.isFinite(cy) &&
      Number.isFinite(w) &&
      Number.isFinite(h) &&
      Number.isFinite(score);
    if (!finite) {
      if (strict) throw new Error("output contains NaN/Inf");
      continue;
    }
    if (score < conf) continue;
    if (allowlist && !allowlist.has(bestC)) continue;
    if (bestC < 0 || bestC >= COCO_CLASSES.length) {
      if (strict) throw new Error(`class_id out of range: ${bestC}`);
      continue;
    }
    if (minArea !== undefined) {
      const area = w * h;
      if (area < minArea) continue;
    }
    boxesArr.push(cx - w * 0.5, cy - h * 0.5, cx + w * 0.5, cy + h * 0.5);
    scoresArr.push(score);
    clsArr.push(bestC);
    idxs.push(i);
  }

  let keepIndices: number[] | null = null;
  if (nms && scoresArr.length > 0) {
    const M0 = scoresArr.length;
    const flatBoxes = new Float32Array(M0 * 4);
    for (let k = 0; k < M0 * 4; k++) flatBoxes[k] = boxesArr[k] as number;
    const flatScores = new Float32Array(scoresArr);
    const flatClasses = new Int32Array(clsArr);
    const kept = classAwareNMS(flatBoxes, flatScores, flatClasses, iouThreshold);
    keepIndices = Array.from(kept);
  }

  const candidateIdxs = keepIndices !== null ? keepIndices : scoresArr.map((_, i) => i);

  const order = candidateIdxs.slice().sort((p, q) => {
    const sd = (scoresArr[q] as number) - (scoresArr[p] as number);
    if (sd !== 0) return sd;
    const cd = (clsArr[p] as number) - (clsArr[q] as number);
    if (cd !== 0) return cd;
    return (idxs[p] as number) - (idxs[q] as number);
  });

  const M = order.length;
  const outBoxes = new Float32Array(M * 4);
  const outScores = new Float32Array(M);
  const outClasses = new Int32Array(M);
  for (let k = 0; k < M; k++) {
    const j = order[k] as number;
    outBoxes[k * 4] = boxesArr[j * 4] as number;
    outBoxes[k * 4 + 1] = boxesArr[j * 4 + 1] as number;
    outBoxes[k * 4 + 2] = boxesArr[j * 4 + 2] as number;
    outBoxes[k * 4 + 3] = boxesArr[j * 4 + 3] as number;
    outScores[k] = scoresArr[j] as number;
    outClasses[k] = clsArr[j] as number;
  }

  if (format === "arrays") {
    return { boxes: outBoxes, scores: outScores, classes: outClasses };
  }
  if (format === "dict") {
    const dets: Detection[] = [];
    for (let k = 0; k < M; k++) {
      const c = outClasses[k] as number;
      dets.push({
        box: [
          outBoxes[k * 4] as number,
          outBoxes[k * 4 + 1] as number,
          outBoxes[k * 4 + 2] as number,
          outBoxes[k * 4 + 3] as number,
        ],
        score: outScores[k] as number,
        class: c,
        label: COCO_CLASSES[c] as string,
      });
    }
    return dets;
  }
  throw new Error(`unknown format: ${format}`);
}
